Group by "pivot-grouping" column

paulrosenzweig opened the pull request:
Fixes #14148

The issue was due to H2's leniency towards ungrouped columns. The `pivot-grouping` column is a constant expression to identify the breakouts used in a row. Not grouping by that works fine for H2 but fails in Postgres. To fix that I added it as a breakout. 

Question: How can I ensure the tests run against all drivers? Or, are they and we never ran `[ci all]`?

paulrosenzweig committed:
group by "pivot-grouping" column

paulrosenzweig committed:
update row order in tests

paulrosenzweig commented on src/metabase/query_processor/pivot.clj:
```diff
@@ -32,7 +32,7 @@
   [query breakout bitmask]
   (-> query
-      (assoc-in [:query :breakout] breakout)
+      (assoc-in [:query :breakout] (conj (seq breakout) [:expression "pivot-grouping"]))
```
`(conj (seq breakout)...` feels wrong. I think one of the values for `breakout` is a vector since it was being appended for one of the breakouts in the test. Is there a better way to append/prepend consistently?

sbelak replied:
`cons` always prepends, but returns a seq which is not ideal (in general the schema calls for a vector and as you've just experienced we don't always gracefully handle other seq types). If you want to consistently append, we usually use either `(concat seq [value])`, or `(conj (vec seq) breakout)`. In general I personally prefer `concat` as it's the most explicit, but in this case explicitly casting to vector is being nice to everyone downstream. 

Having said all that, if this is only to make the Testing Gods happy, I'd be tempted to just change the tests and sneak in something along the lines of `(update :breakout sort)`, rather than complicating things here.

paulrosenzweig committed:
fix tests

paulrosenzweig committed:
fix assertion for successful response

paulrosenzweig committed:
set up question without UI

paulrosenzweig committed:
eslint

paulrosenzweig committed:
random attempt to fix test before I set things up locally

paulrosenzweig committed:
switch to other table :shrug:

sbelak reviewed (changes requested):
All tests should be wrapped in `(mt/test-drivers (mt/normal-drivers-with-feature :expressions)` else things will break for drivers that don't support expressions (incidentally how is this -- driver not supporting expressions -- accounted for in the UI and in normal usage). 

Aside from that, looks good. 

`[ci all]` forces the tests to be run against all drivers.

paulrosenzweig commented:
> incidentally how is this -- driver not supporting expressions -- accounted for in the UI and in normal usage

😬 it's not. If there's no clear workaround on the server, we can disable pivot tables for those dbs.

paulrosenzweig committed:
pr feedback [ci all]

paulrosenzweig committed:
whoops! keyboard macro gone wrong [ci all]

paulrosenzweig committed:
forgot to update a test [ci all]

paulrosenzweig committed:
another test [ci all]

paulrosenzweig committed:
put number in right place [ci all]

sbelak commented:
> > incidentally how is this -- driver not supporting expressions -- accounted for in the UI and in normal usage
> 
> 😬 it's not. If there's no clear workaround on the server, we can disable pivot tables for those dbs.

I don't think there's a clean workaround. We perhaps picked the wrong abstraction level and now there's no way for a driver to overwrite the pivot computation implementation. Probably best disable it for now and open a followup issue.

Commit:
Fixing up the errors on PostgreSQL (and other SQL databases)

Also ensuring tests run against drivers that support both expressions
and left-joins

[ci all]

camsaul commented:
You don't need to use `[ci all]` anymore -- CI will do the right thing automatically now

camsaul reviewed (approved):
LGTM

robdaemon commented:
@camsaul sweet, I'm working on resolving the driver tests right now, otherwise they'll haunt us (they're not specific to this change, but because of the new `sample-dataset`)

Commit:
fixing test failures

Commit:
On Oracle and Snowflake, we need to add IDs *and* insert in smaller
batches

robdaemon commented:
so I had to make a few changes to the Snowflake driver for this, and then go into the Snowflake web UI and drop the broken `v3_sample_dataset` database so it could be reinserted.

Redshift seems to be timing out, so I'll see if there's another way to pre-load the data into Redshift.

Commit:
Do not run these tests on Redshift

it takes more than 10 minutes just to insert the sample-dataset into
Redshift, so do not run these tests against Redshift. All other
databases in the `applicable-drivers` def are valid though.

sbelak reviewed (approved):
Looks good
